controller/gateway/controller.go

```diff
@@ -309,11 +309,16 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	controlplane := r.provisionControlPlane(ctx, logger, gwc.GatewayClass, &gateway, gatewayConfig, dataplane, ingressServices[0], adminServices[0])
 
 	// Set the ControlPlaneReady Condition to False. This happens only if:
-	// * the new status is false and there was no ControlPlaneReady condition in the cgatewafy
+	// * the new status is false and there was no ControlPlaneReady condition in the gateway
 	// * the new status is false and the previous status was true
-	if !k8sutils.IsConditionTrue(ControlPlaneReadyType, gwConditionAware) {
-		condition, found := k8sutils.GetCondition(ControlPlaneReadyType, oldGwConditionsAware)
-		if !found || condition.Status == metav1.ConditionTrue {
+	if condition, found := k8sutils.GetCondition(ControlPlaneReadyType, gwConditionAware); found && condition.Status != metav1.ConditionTrue {
+		if condition.Reason == string(consts.UnableToProvisionReason) {
+			log.Debug(logger, "unable to provision controlplane, requeueing", gateway)
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		conditionOld, foundOld := k8sutils.GetCondition(ControlPlaneReadyType, oldGwConditionsAware)
+		if !foundOld || conditionOld.Status == metav1.ConditionTrue {
```

Comment:
I'm pretty scared by these changes as the Gateway status historically has been a delicate point to touch. What's the root cause of this change here? Any bug we need to fix?

Reply:
The only change in here is in the lines 315-318 which requeues when the CP failed to get provisioned. This could happen (as described in the PR description) when e.g. webhook is not available (this can happen when operator is upgraded).

Without this requeue the reconciliation gets stuck and the only indication of such a situation is a condition on the Gateway.